rp2040-hal/src/pio.rs

```diff
@@ -531,17 +527,25 @@ impl<SM: ValidStateMachine, State> StateMachine<SM, State> {
     #[deprecated(note = "Renamed to exec_instruction")]
     ///Execute the instruction immediately.
     pub fn set_instruction(&mut self, instruction: u16) {
+        let instruction =
+            Instruction::decode(instruction, self.program.side_set).expect("Invalid instruction");
         self.exec_instruction(instruction);
     }
 
     /// Execute the instruction immediately.
     ///
-    /// While this is allowed even when the state machine is running, the datasheet says:
-    /// > If EXEC instructions are used, instructions written to INSTR must not stall.
-    /// It's unclear what happens if this is violated.
-    pub fn exec_instruction(&mut self, instruction: u16) {
-        // TODO: clarify what happens if the instruction stalls.
-        self.sm.exec_instruction(instruction);
+    /// If an instruction written to INSTR stalls, it is stored in the same instruction latch used
+    /// by OUT EXEC and MOV EXEC, and will overwrite an in-progress instruction there. If EXEC
+    /// instructions are used, instructions written to INSTR must not stall.
+    pub fn exec_instruction(&mut self, instruction: Instruction) {
+        let instruction = instruction.encode(self.program.side_set);
+
+        critical_section::with(|_| {
```

Comment:
I don't like all those added critical sections.
Most of those instances are already guarded from concurrent accesses by the fact that the functions take a mutable reference to the state machine.

The only exception I found is `Tx::drain_fifo(&mut self)`. `Tx` doesn't contain a reference to the state machine, and the write access is realized with help of an unsafe access in `Tx::register_block(&self)`.
What about moving `drain_fifo` to the state machine, or alternatively adding a `&mut StateMachine` parameter, to guarantee exclusive access. That would make the critical sections unnecessary.

BTW, this safety comment is very sus, as the same function appears in both `Tx` and `Rx`:
```
    fn register_block(&self) -> &pac::pio0::RegisterBlock {
        // Safety: The register is unique to this Tx instance.
        unsafe { &*self.block }
    }
```

Reply:
Yes, that's the only reason we need those critical sections. I was tempted to make those calls non-mut but I didn't like that.

It seemed weird to have the drain fifo for tx on the SM without one for Rx, but I'm not against it.
Do you think we should go this way to get rid of the critical sections ?

Reply:
Yes, I very much prefer the solution without the critical sections.
While the additional overhead is probably not significant for those functions, critical sections are a global resource: They could affect other users of critical sections, and add noticeable interrupt latency.
I think any critical section in library code should be avoided if possible, and if really needed, they should be limited to code which usually only runs once during initialization, or be as short as possible (eg. a read/modify/write operation).